components/file_system/src/iosnoop/biosnoop.rs

```diff
@@ -0,0 +1,332 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+use super::metrics::*;
+use super::IOStats;
+use crate::IOType;
+
+use collections::HashMap;
+use std::collections::VecDeque;
+use std::ffi::CString;
+use std::ptr;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Mutex;
+
+use bcc::{table::Table, Kprobe, BPF};
+use crossbeam_utils::CachePadded;
+
+const MAX_THREAD_IDX: usize = 128;
+
+static mut BPF_TABLE: Option<(BPF, Table, Table)> = None;
+// This array records the io type for every thread. The address of this array
+// will be passed into BPF, so BPF code can get io type for specific thread
+// without an extra syscall.
+// It should be a thread local variable, but the address of thread local is not
+// reliable. So define a global array and let each thread writes on a specific
+// element. Thus there is no contention for the element and use padding to avoid
+// false sharing.
+static mut IO_TYPE_ARRAY: [CachePadded<IOType>; MAX_THREAD_IDX] =
+    [CachePadded::new(IOType::Other); MAX_THREAD_IDX];
+
+// The index of the element of IO_TYPE_ARRAY for this thread to access.
+thread_local! {
+    static IDX: IdxWrapper = unsafe {
+        let idx = IDX_ALLOCATOR.allocate();
+        if let Some((_, _, t)) = BPF_TABLE.as_mut() {
+            let tid = nix::unistd::gettid().as_raw() as u32;
+            let ptr : *const *const _ = &IO_TYPE_ARRAY.as_ptr().add(idx.0);
+            t.set(&mut tid.to_ne_bytes(), std::slice::from_raw_parts_mut(ptr as *mut u8, std::mem::size_of::<*const IOType>())).unwrap();
```

Comment:
Is it only called when thread initialize? We need to set mixed IO type for, e.g. the unified thread pool.

Reply:
it's lazy initlialized when `get_io_type()` or `set_io_type()` is called for the first time. For mixed IO, we need to call `set_io_type()` explicitly everytime switch the task.